ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -349,42 +348,31 @@ class SimpleOperationTracker implements OperationTracker {
         remoteOfflineReplicas.forEach(this::addToEndOfPool);
       }
     }
-    totalReplicaCount = replicaPool.size();
-    originatingDcOfflineReplicaCount =
-        getReplicasByState(originatingDcName, EnumSet.of(ReplicaState.OFFLINE)).values().size();
-    originatingDcEligibleReplicaCount = replicaPool.stream()
-        .filter(replicaId -> replicaId.getDataNodeId().getDatacenterName().equals(this.originatingDcName))
-        .collect(Collectors.toList())
-        .size();
-    originatingDcTotalReplicaCount = allReplicas.stream()
-        .filter(replicaId -> replicaId.getDataNodeId().getDatacenterName().equals(this.originatingDcName))
-        .collect(Collectors.toList())
-        .size();
+    eligibleReplicaCount = replicaPool.size();
 
     // MockPartitionId.getReplicaIds() is returning a shared reference which may cause race condition.
     // Please report the test failure if you run into this exception.
     Supplier<IllegalArgumentException> notEnoughReplicasException = () -> new IllegalArgumentException(
         generateErrorMessage(partitionId, examinedReplicas, replicaPool, backupReplicasToCheck, downReplicasToCheck,
             routerOperation));
-    if (totalReplicaCount < getSuccessTarget()) {
+    if (eligibleReplicaCount < getSuccessTarget()) {
       throw notEnoughReplicasException.get();
     }
 
-    int numActiveReplicasInOriginatingDc =
-        getEligibleReplicas(originatingDcName, EnumSet.of(ReplicaState.STANDBY, ReplicaState.LEADER)).size();
-    if (routerConfig.routerOperationTrackerTerminateOnNotFoundEnabled
-        && numActiveReplicasInOriginatingDc >= routerConfig.routerPutSuccessTarget) {
-      // There are two conditions to meet in order to use this feature
-      // 1. TerminateOnNotFound has to be enabled.
-      // 2. We need enough active replicas in the originating DC
-      // How many active replicas is considered as enough? We have to consider some extreme cases.
-      // When the put success target is 2, we at least need 2 replicas in active state so we know the replicas are up to
-      // date. If we have only one replica in active state, we might run into a scenario that we are getting NotFound
-      // from the other two inactive replicas and we should not ignore the active one.
-      originatingDcNotFoundFailureThreshold = originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1;
-    } else {
-      originatingDcNotFoundFailureThreshold = 0;
+    originatingDcTotalReplicaCount = (int) allReplicas.stream()
+        .filter(replicaId -> replicaId.getDataNodeId().getDatacenterName().equals(this.originatingDcName))
```

Comment:
Sure. Since it seems like .equals(null) would return false, `originatingDcTotalReplicaCount` would be 0. Added check that `originatingDcNotFoundFailureThreshold` should be > 0 in `hasFailedNotFound`.